Add blog entry Crimson balance CPU (#797)
ceph/ceph.io: src/en/news/blog/2025/crimson-balance-cpu-part1/index.md

decided only show the representative 8 OSDs and 5 I/O reactors.  

<!-- A .pdf report with full details is available here.  -->

We used a single node cluster, with the following hardware and system configuration:

- CPU: 2 x Intel(R) Xeon(R) Platinum 8276M CPU @ 2.20GH (56 cores each)
- Memory: 394 GiB
- Drives: 8 x 93.1 TB NVMe
- OS: Centos 9.0 on kernel 5.14.0-511.el9.x86_64
- Ceph: version b2a220 (bb2a2208867d7bce58b9697570c83d995a1c5976) squid (dev)
- podman version 5.2.2.

We build Ceph with the following options:

```bash
# ./do_cmake.sh -DWITH_SEASTAR=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

At a high level, we initiate the performance tests as follows:

```bash
/root/bin/run_balanced_crimson.sh -t cyan
```
This will run the test plan for the Cyanstore object backend, producing data for
response curves over the three CPU allocation strategies. The argument
'`-t`' is used to specify the object storage backend: `cyan`, `sea` and `blue` for
Cyanstore, Seastore and Bluestore, respectively.

```bash
crimson_be_table["cyan"]="--cyanstore"
crimson_be_table["sea"]="--seastore --seastore-devs ${STORE_DEVS}"
crimson_be_table["blue"]="--bluestore --bluestore-devs ${STORE_DEVS}"
```

> Once the execution completes, the results are archived in .zip files according

Reviewer: nit: 
```suggestion
Once the execution completes, the results are archived in .zip files according
to the workloads and saved in the output directory which can be specified with the `-d` option. (`/tmp' by default).
```

    from the OSD process and the FIO client process via top (30 samples over 5
    minutes).
    - `cyan_8osd_5reactor_8fio_bal_osd_rc_1procs_randread_diskstat.json`: the
    diskstat output. A sample is taken before and after the test, the .json
    contains the differences.
    - `cyan_8osd_5reactor_8fio_bal_osd_rc_1procs_randread_top.json`: the output
    from top, parsed via jc to produce a .json. **Note**: jc does not yet
    support individual CPU core utilisation, so we have to rely on the overall
    CPU utilisation (per thread).
    - `new_cluster_dump.json`: the output from `ceph tell ${osd_i}
    dump_metrics` command, which contains the individual OSD performance
    metrics.
    - `FIO_cyan_8osd_5reactor_8fio_bal_osd_rc_1procs_randread_top_cpu.plot`:
    the plot mechanically generated from the top output, showing the CPU
    utilisation over time for the FIO client.
    - `OSD_cyan_8osd_5reactor_8fio_bal_osd_rc_1procs_randread_top_mem.plot`:
    the plot mechanically generated from the top output, showing the MEM
    utilisation over time for the OSD process.

</details>

To produce the post-processing and side-by-side comparison, the following script

> is run: the arguments are the input directory that contains the runs we want to

Reviewer: I think we can move the description of the arguments below the command to run the script.

compare, the type of object store backend, and the output .md to produce.
```bash
# /root/bin/pp_balanced_cpu_cmp.sh -d /tmp/_seastore_8osd_5_6_reactor_8fio_rc_cmp \
   -t sea -o seastore_8osd_5vs6_reactor_8fio_cpu_cmp.md
```

We will show the comparisons produced in the next section.

We end this section by looking behind the curtains of the above scripts,

> showing details on precondition the drives, creation of the cluster, execution

Reviewer: ```suggestion
showing details on preconditioning the drives, creation of the cluster, execution
```

> of FIO and the metrics collected.

Reviewer: These can be written as bullet points for more clarity

    We use a single job per RBD volume (but this could also be variable if required).

```bash
# Option -w (WORKLOAD) is used as index for these:
declare -A m_s_iodepth=( [hockey]="1 2 4 8 16 24 32 40 52 64"  ...)
declare -A m_s_numjobs=( [hockey]="1" ... )
```

- As a preliminary step, we prime the volume(s) with a write workload. This is done
before the actual performance tests are run. We exercise a client per volume,
so the execution is concurrent.

```bash
# Prime the volume(s) with a write workloads
   RBD_NAME=fio_test_$i RBD_SIZE="64k" fio ${FIO_JOBS}rbd_prime.fio 2>&1 >/dev/null & 
   echo "== priming $RBD_NAME ==";
 ...
   wait;
```

- In the main loop, we iterate over the number of I/O depths, and we run the
FIO command for each of the predefined workloads. We also take a snapshot of
the diskstats before and after the test, and we calculate the difference. We
collect the PIDs of the FIO process as well as the OSD processes, which would
be used to monitor their resource utilisation. We have an heuristic in place to
exit early the loops if the standard deviation of the latency disperses too
much from the median.

> **IMPORTANT**: the attentive reader would notice the use the `taskset` command

Reviewer: ```suggestion
**IMPORTANT**: the attentive reader would notice the use of the `taskset` command
```

...
      wait;
      # Measure the diskstats after the completion of FIO
      jc --pretty /proc/diskstats | python3 /root/bin/diskstat_diff.py -a ${DISK_STAT}

      # Exit the loops if the latency disperses too much from the median
      if [ "$RESPONSE_CURVE" = true ] && [ "$RC_SKIP_HEURISTIC" = false ]; then
        mop=${mode[${WORKLOAD}]}
        covar=$(jq ".jobs | .[] | .${mop}.clat_ns.stddev/.${mop}.clat_ns.mean < 0.5 and \
          .${mop}.clat_ns.mean/1000000 < ${MAX_LATENCY}" fio_${TEST_NAME}.json)
                  if [ "$covar" != "true" ]; then
                    echo "== Latency std dev too high, exiting loops =="
                    break 2
                  fi
      fi
    done # loop IODEPTH
  done # loop NUM_JOBS 

```
The basic monitoring routine is shown below, which is executed concurrently as FIO progresses. 
```bash
fun_measure() {
  local PID=$1 #comma separated list of pids
  local TEST_NAME=$2
  local TEST_TOP_OUT_LIST=$3

  top -b -H -1 -p "${PID}" -n ${NUM_SAMPLES} >> ${TEST_NAME}_top.out
  echo "${TEST_NAME}_top.out" >> ${TEST_TOP_OUT_LIST}
}
```
</details>

We have written a custom profile for top, so we get information about the

> parent process id, last CPU the thread was executed, etc. (which are not

Reviewer: last CPU the thread was executed on?

|:---------:|:---------:|
|![cyan_8osd_5reactor_8fio_randread_fio_cpu](images/cyan_8osd_5reactor_8fio_randread_fio_cpu.png) | ![cyan_8osd_5reactor_8fio_randread_fio_mem](images/cyan_8osd_5reactor_8fio_randread_fio_mem.png) |

</details>

#### Seastore
![sea_8osd_5reactor_8fio_randread_bal_vs_unbal_iops_vs_lat](images/sea_8osd_5reactor_8fio_randread_bal_vs_unbal_iops_vs_lat.png)

<details>
<summary>Click to see CPU and MEM utilisation.</summary>

- It is a good sign to see a constant memory utilisation for the OSD process
across all the iodepths. The NUMA socket strategy has only a .6% higher than
the other strategies, which is not significant to cause concern.

| _OSD CPU_ | _OSD MEM_ |
|:---------:|:---------:|
| ![sea_8osd_5reactor_8fio_randread_osd_cpu](images/sea_8osd_5reactor_8fio_randread_osd_cpu.png) | ![sea_8osd_5reactor_8fio_randread_osd_mem](images/sea_8osd_5reactor_8fio_randread_osd_mem.png)|

| _FIO CPU_ | _FIO MEM_ |
|:---------:|:---------:|
|![sea_8osd_5reactor_8fio_randread_fio_cpu](images/sea_8osd_5reactor_8fio_randread_fio_cpu.png) |![sea_8osd_5reactor_8fio_randread_fio_mem](images/sea_8osd_5reactor_8fio_randread_fio_mem.png)|

</details>

#### Bluestore

> - Interestingly, the NUMA socket balanced CPU allocation strategy seem to have a slight advantage over the other CPU allocation strategies for Bluestore.

Reviewer: ```suggestion
- Interestingly, the NUMA socket balanced CPU allocation strategy seemed to have a slight advantage over the other CPU allocation strategies for Bluestore.
```

| _FIO CPU_ | _FIO MEM_ |
|:---------:|:---------:|
|![sea_8osd_5reactor_8fio_seqread_fio_cpu](images/sea_8osd_5reactor_8fio_seqread_fio_cpu.png)|![sea_8osd_5reactor_8fio_seqread_fio_mem](images/sea_8osd_5reactor_8fio_seqread_fio_mem.png)|
</details>

#### Bluestore
![blue_8osd_5reactor_160at_8fio_seqread_bal_vs_unbal_iops_vs_lat](images/blue_8osd_5reactor_160at_8fio_seqread_bal_vs_unbal_iops_vs_lat.png)
<details>
<summary>Click to see CPU and MEM utilisation.</summary>

| _OSD CPU_ | _OSD MEM_ |
|:---------:|:---------:|
|![blue_8osd_5reactor_160at_8fio_seqread_osd_cpu](images/blue_8osd_5reactor_160at_8fio_seqread_osd_cpu.png)|![blue_8osd_5reactor_160at_8fio_seqread_osd_mem](images/blue_8osd_5reactor_160at_8fio_seqread_osd_mem.png)|

| _FIO CPU_ | _FIO MEM_ |
|:---------:|:---------:|
|![blue_8osd_5reactor_160at_8fio_seqread_fio_cpu](images/blue_8osd_5reactor_160at_8fio_seqread_fio_cpu.png)|![blue_8osd_5reactor_160at_8fio_seqread_fio_mem](images/blue_8osd_5reactor_160at_8fio_seqread_fio_mem.png)|
</details>

### Sequential 64K write

> - Is hard to explain the initial spike in the CPU utilisation for the three CPU

Reviewer: ```suggestion
- It is hard to explain the initial spike in the CPU utilisation for the three CPU
```

> allocation strategies, across the three object backend. We might need to run a

Reviewer: ```suggestion
allocation strategies, across the three object store backends. We might need to run a
```
